cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -3319,9 +3324,401 @@ struct GroupedListFactory {
+        }
+      }
+      return Status::OK();
+    };
+
+    // Call compute_take_indices with the optimal integer width
+    if (values.length <= static_cast<int64_t>(std::numeric_limits<uint8_t>::max())) {
+      RETURN_NOT_OK(compute_take_indices(uint8_t{}));
+    } else if (values.length <=
+               static_cast<int64_t>(std::numeric_limits<uint16_t>::max())) {
+      RETURN_NOT_OK(compute_take_indices(uint16_t{}));
+    } else if (values.length <=
+               static_cast<int64_t>(std::numeric_limits<uint32_t>::max())) {
+      RETURN_NOT_OK(compute_take_indices(uint32_t{}));
+    } else {
+      RETURN_NOT_OK(compute_take_indices(uint64_t{}));
+    }
+
+    // Use take_indices to update column #key
+    auto values_data = values.ToArrayData();
+    auto indices_data = ArrayData::Make(
+        take_index_type, num_groups_, {std::move(take_bitmap), std::move(take_indices)});
+    ARROW_ASSIGN_OR_RAISE(Datum grouped_column,
+                          Take(values_data, indices_data, TakeOptions::Defaults(), ctx_));
+    return MergeColumn(&columns_[key], grouped_column.make_array());
+  }
+
+  Status Resize(int64_t new_num_groups) {
+    if (new_num_groups > std::numeric_limits<int32_t>::max()) {
+      return Status::NotImplemented("Pivot with more 2**31 groups");
+    }
+    return ResizeColumns(new_num_groups);
+  }
+
+  Status Merge(GroupedPivotAccumulator&& other, const ArrayData& group_id_mapping) {
+    // To merge `other` into `*this`, we simply merge their respective columns.
+    // However, we must first transpose `other`'s rows using `group_id_mapping`.
+    // This is a logical "scatter" operation.
+    //
+    // Since `scatter(indices)` is implemented as `take(inverse_permutation(indices))`,
+    // we can save time by computing `inverse_permutation(indices)` once for all
+    // columns.
+
+    // Scatter/InversePermutation only accept signed indices. We checked
+    // in Resize() above that we were inside the limites for int32.
+    auto scatter_indices = group_id_mapping.Copy();
+    scatter_indices->type = int32();
+    std::shared_ptr<DataType> take_indices_type;
+    if (num_groups_ - 1 <= std::numeric_limits<int8_t>::max()) {
+      take_indices_type = int8();
+    } else if (num_groups_ - 1 <= std::numeric_limits<int16_t>::max()) {
+      take_indices_type = int16();
+    } else {
+      DCHECK_GE(num_groups_ - 1, std::numeric_limits<int32_t>::max());
+      take_indices_type = int32();
+    }
+    InversePermutationOptions options(/*max_index=*/num_groups_ - 1, take_indices_type);
+    ARROW_ASSIGN_OR_RAISE(auto take_indices,
+                          InversePermutation(scatter_indices, options, ctx_));
+    auto scatter_column =
+        [&](const std::shared_ptr<Array>& column) -> Result<std::shared_ptr<Array>> {
+      ARROW_ASSIGN_OR_RAISE(auto scattered,
+                            CallFunction("take", {column, take_indices}, &options, ctx_));
```

Comment:
Why is this not using `Take(...)` like the other two invocations?

Reply:
I don't think there was a precise reason, let me try to make this more consistent :)

Reply:
Ahah, and this was actually hiding a bug. It was passing a `InversePermutationOptions` instead of `TakeOptions`.

Reply:
Oops, missed that actually.